### PR TITLE
rpcserver: Address validation in SendCoins according to network

### LINKED ISSUE #2677: SendCoins does not check for mainnet or testnet
### Background

The gRPC SendCoin command does not check if the destination address belongs to testnet or mainnet.
It is not a big deal, as any developer can check this before calling the command, but still I think it would be nice to have that extra security.
Note: I only tested sending testcoins to mainnet addresses, not the other way arround.

### Your environment

LND: 0.5.1-beta commit=v0.5.1-beta-263-g552a9b7190e14269f612f7fda401172d6e81f147
operating system: Linux Manjaro

### Steps to reproduce

1. Configure your LND to use testnet
2. Use SendCoin gRPC command to send testnet coins to a mainnet address

### Expected behaviour

It should throw an error: Wrong network

### Actual behaviour

It works and the coins are gone

### PR DESCRIPTION
**Pull request by ccdle12**
fixes #2677

This PR updates `SendCoins` in `rpcserver.go` to check that the receiving address of an on chain transaction is a valid address according to the current network of the user.

### CODE REVIEW

**Comment by michaelWuensch** [comment]
Hey, thanks for fixing the issue I reported! Had a quick glance at your code.
I think for your second thest the error should be:

t.Fatalf("there should be an error since we are attempting to send coins to an **mainnet** address")

**Comment by ccdle12** [comment]
> Hey, thanks for fixing the issue I reported! Had a quick glance at your code.
> I think for your second test the error should be:
> 
> t.Fatalf("there should be an error since we are attempting to send coins to an **mainnet** address")

Thanks for that catch!

**Comment by molxyz** [comment]
tACK
I just tested this PR by sending test coins to a few mainnet bitcoin addresses, it works!
Without this PR, sending coin to a mainnet bitcoin address looks like this: 
```
$ tlncli sendcoins 1KFHE7w8BhaENAswwryaoccDb6qcT6DbYY 50000
{
    "txid": "eede68e0f976a47750fa6f83b5341c87199d536464277fee5654e45960822922"
}
```
But here's the results after sending test coins to mainnet bitcoin addresses:
```
~$ tlncli sendcoins 1KFHE7w8BhaENAswwryaoccDb6qcT6DbYY 50000
[lncli] rpc error: code = Unknown desc = address: 1KFHE7w8BhaENAswwryaoccDb6qcT6DbYY is not valid for this network: testnet3
~$ tlncli sendcoins bc1qjl8uwezzlech723lpnyuza0h2cdkvxvh54v3dn 300000
[lncli] rpc error: code = Unknown desc = address: bc1qjl8uwezzlech723lpnyuza0h2cdkvxvh54v3dn is not valid for this network: testnet3
~$ tlncli sendcoins 3JXuYz6gcNV7nMNaNx2jb7zqj83KBsn3UZ 16777215
[lncli] rpc error: code = Unknown desc = address: 3JXuYz6gcNV7nMNaNx2jb7zqj83KBsn3UZ is not valid for this network: testnet3
```

**Comment by molxyz** [comment]
OK I just tested this PR on my mainnet node by sending to testnet addresses, and here's the result:
```
~$ lncli sendcoins mukF8KqFcqnpm3GiU9SjMAUfFLK6zyX3Rq 10000 --sat_per_byte=12
[lncli] rpc error: code = Unknown desc = address: mukF8KqFcqnpm3GiU9SjMAUfFLK6zyX3Rq is not valid for this network: mainnet
~$ lncli sendcoins tb1qcel2wzczlvpxv5c7jhmrhncvw00rmqhwlt30zu 10000 --sat_per_byte=10
[lncli] rpc error: code = Unknown desc = address: tb1qcel2wzczlvpxv5c7jhmrhncvw00rmqhwlt30zu is not valid for this network: mainnet
~$ lncli sendcoins 2Mzc1mniVDUksTsW7zvX2pwVwWPd6rEmsSq 10000
[lncli] rpc error: code = Unknown desc = address: 2Mzc1mniVDUksTsW7zvX2pwVwWPd6rEmsSq is not valid for this network: mainnet
```
So it seems this patch works as it's intended. 
@ccdle12  Thank you!

File: rpcserver.go [hunk 1]
```diff
@@ -792,6 +792,22 @@ func (r *rpcServer) SendCoins(ctx context.Context,
...
+		in.Addr, activeNetParams.Params,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make the check on the decoded address according to the active network.
+	if !addr.IsForNet(activeNetParams.Params) {
+		return nil, fmt.Errorf("address: %v is not valid for this network: %v",
+			addr.String(), activeNetParams.Params.Name,
```
Comment (wpaulino) [inline rpcserver.go]: We should normalize these names as regtest, for example, will return "testnet3".

**Comment by ccdle12** [comment]
Thanks for the review @wpaulino! , I'll get started on the comments

File: lnd_test.go [hunk 2]
```diff
@@ -5059,6 +5059,66 @@ func testPrivateChannels(net *lntest.NetworkHarness, t *harnessTest) {
 	closeChannelAndAssert(ctxt, t, net, carol, chanPointPrivate, false)
 }
 
+// testSendCoinsAddressValidation tests that when attempting to send coins on
+// chain, the address is validated according to the current network.
+func testSendCoinsAddressValidation(net *lntest.NetworkHarness, t *harnessTest) {
```
[the remark this answers is not in the document]
Comment (ccdle12) [reply lnd_test.go]: fixed!

File: lnd_test.go [hunk 3]
```diff
@@ -5059,6 +5059,66 @@ func testPrivateChannels(net *lntest.NetworkHarness, t *harnessTest) {
...
+	// Attempt to send coins to a testnet address, should fail since we are in
+	// simnet.
+	ctxt, _ = context.WithTimeout(ctxb, defaultTimeout)
+	sendCoinReq = &lnrpc.SendCoinsRequest{
+		Addr:   "tb1qfc8fusa98jx8uvnhzavxccqlzvg749tvjw82tg",
+		Amount: 500000,
+	}
+	_, err = net.Alice.SendCoins(ctxt, sendCoinReq)
+	if err == nil {
+		t.Fatalf("there should be an error since we are attempting to send " +
```
[the remark this answers is not in the document]
Comment (ccdle12) [reply lnd_test.go]: fixed!

File: rpcserver.go [hunk 4]
```diff
@@ -792,6 +792,22 @@ func (r *rpcServer) SendCoins(ctx context.Context,
 		in.Addr, btcutil.Amount(in.Amount), int64(feePerKw),
 		in.SendAll)
 
+	// Decode the address receiving the coins, we need to check whether the
+	// address is valid for this network.
+	addr, err := btcutil.DecodeAddress(
```
[the remark this answers is not in the document]
Comment (ccdle12) [reply rpcserver.go]: fixed!

File: rpcserver.go (hunk 1 above)
Comment (ccdle12) [reply rpcserver.go]: I saw the other PR, it has a function:

```
// activeNetParams returns a string representation of the current active
// network.
func activeNetName() (string, error) {
	switch {
	case cfg.Bitcoin.MainNet || cfg.Litecoin.MainNet:
		return "mainnet", nil
	case cfg.Bitcoin.TestNet3 || cfg.Litecoin.TestNet3:
		return "testnet", nil
	case cfg.Bitcoin.SimNet || cfg.Litecoin.SimNet:
		return "simnet", nil
	case cfg.Bitcoin.RegTest || cfg.Litecoin.RegTest:
		return "regest", nil
	default:
		return "", fmt.Errorf("unknown network for addr")
	}
}
```

Is this what you meant = normalize an input from ("testnet3") -> ("testnet")?

Comment (wpaulino) [reply rpcserver.go]: Yep!

Comment (ccdle12) [reply rpcserver.go]: @wpaulino Thanks! I basically just reapplied roasbeefs changes including refactoring duplicate use of DecodeAddress

File: rpcserver.go [hunk 5]
```diff
@@ -799,6 +799,23 @@ func (r *rpcServer) ListUnspent(ctx context.Context,
...
+func activeNetName() (string, error) {
+	switch {
+	case cfg.Bitcoin.MainNet || cfg.Litecoin.MainNet:
+		return "mainnet", nil
+	case cfg.Bitcoin.TestNet3 || cfg.Litecoin.TestNet3:
+		return "testnet", nil
+	case cfg.Bitcoin.SimNet || cfg.Litecoin.SimNet:
+		return "simnet", nil
+	case cfg.Bitcoin.RegTest || cfg.Litecoin.RegTest:
+		return "regest", nil
```
Comment (dannypaz) [inline rpcserver.go]: was casually browsing for issues on lnd and noticed the typo here. `regest`

Comment (ccdle12) [reply rpcserver.go]: thanks for the catch! fixed.

**Comment by wpaulino** [comment]
@ccdle12 LGTM! Could you rebase and squash the fixup commits?

**Comment by ccdle12** [comment]
> @ccdle12 LGTM! Could you rebase and squash the fixup commits?

@wpaulino Thanks for the review! rebased and squashed

File: rpcserver.go [hunk 6]
```diff
@@ -858,6 +858,23 @@ func (r *rpcServer) EstimateFee(ctx context.Context,
 	return resp, nil
 }
 
+// activeNetParams returns a string representation of the currect active
+// network.
+func activeNetName() (string, error) {
```
Comment (halseth) [inline rpcserver.go]: since this is only used for logging, I don't think it is worth it.

File: rpcserver.go [hunk 7]
```diff
@@ -880,6 +897,25 @@ func (r *rpcServer) SendCoins(ctx context.Context,
...
+	// Decode the address receiving the coins, we need to check whether the
+	// address is valid for this network.
+	targetAddr, err := btcutil.DecodeAddress(in.Addr, activeNetParams.Params)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make the check on the decoded address according to the active network.
+	if !targetAddr.IsForNet(activeNetParams.Params) {
+		netName, err := activeNetName()
```
Comment (halseth) [inline rpcserver.go]: just use `activeNetParams.Params.Name` instead.

**Comment by Roasbeef** [comment]
Committed as abfbdf6aec3bbf63a1f13e5706ee983efbfb4674

**Pull request closed without merging**